python/pylibcudf/tests/test_lists.py

```diff
@@ -161,28 +160,34 @@ def test_reverse(list_column):
     assert_column_eq(expect, got)
 
 
-# https://github.com/rapidsai/cudf/issues/19346
-@pytest.mark.flaky(reruns=5)
-def test_segmented_gather(test_data):
+@pytest.mark.parametrize(
+    "bounds_policy",
+    [
+        plc.copying.OutOfBoundsPolicy.DONT_CHECK,
+        plc.copying.OutOfBoundsPolicy.NULLIFY,
+    ],
+    ids=["DONT_CHECK", "NULLIFY"],
+)
+def test_segmented_gather(
+    test_data: list[list[list[int]]],
+    bounds_policy: plc.copying.OutOfBoundsPolicy,
+) -> None:
     list_column1, list_column2 = test_data[0]
 
-    plc_column1 = plc.Column.from_arrow(pa.array(list_column1))
-    plc_column2 = plc.Column.from_arrow(pa.array(list_column2))
-
-    got = plc.lists.segmented_gather(plc_column2, plc_column1)
-
-    expect = pa.array([[8, 9], [14], [0], [0, 0]])
-
-    assert_column_eq(expect, got)
-
-
-def test_extract_list_element_scalar(list_column):
-    plc_column = plc.Column.from_arrow(pa.array(list_column))
+    gather_map_list = plc.Column.from_arrow(pa.array(list_column1))
+    input_column = plc.Column.from_arrow(pa.array(list_column2))
 
-    got = plc.lists.extract_list_element(plc_column, 0)
-    expect = pc.list_element(list_column, 0)
+    got = plc.lists.segmented_gather(
+        input_column, gather_map_list, bounds_policy
+    )
 
-    assert_column_eq(expect, got)
+    if bounds_policy == plc.copying.OutOfBoundsPolicy.DONT_CHECK:
+        # Everything except [0][0] is out-of-bounds indexing.
+        assert got.size() == 4
+        assert plc.interop.to_arrow(got)[0][0].as_py() == 8
```

Comment:
It's expected that `DONT_CHECK` is only used when the inputs are known to be in-bounds. Otherwise I think this could trigger out-of-bounds reads which can cause CUDA failures. We don't want to trigger the UB described in the docstring.

I would propose that we modify the test in the following way:
- Test `DONT_CHECK` with in-bounds gather maps but _not_ out-of-bounds gather maps
- Test `NULLIFY` with in-bounds gather maps and out-of-bounds gather maps (to test nullification)

Reply:
> If indices in gather_map_list are outside the range [-n, n), where n is the number of elements in corresponding row of the source column, the behavior is as follows:
> 1. If bounds_policy is set to DONT_CHECK, **the behavior is undefined**.
> 2. If bounds_policy is set to NULLIFY, the corresponding element in the list row is set to null in the output column.

(Emphasis added).

In C++, undefined behavior means anything can happen, including crashes of the program. This phrasing may be unfamiliar to Python users, so clarifying this is a good idea.

Reply:
Thanks for the clarification. Hopefully we're OK now:

- https://github.com/rapidsai/cudf/pull/19411/commits/ad9269e66cc333b97653886b83138e5d4c43fbd7 updates the tests by splitting it in two:

  1. a test with all in-bounds indices that tests both `DONT_CHECK` and `NULLIFY`.
  2. a test with some out-of-bounds indices with just `NULLIFY`

i.e. we don't perform any out-of-bounds indexing with `DONT_CHECK`

- https://github.com/rapidsai/cudf/pull/19411/commits/448c069b755674ce41583e37b7d73cd0c147a7ae clarifies the docs in pylibcudf's wrapper about what's covered by the scope of "undefined".